### PR TITLE
Add v1/me/business event endpoint

### PR DESCRIPTION
**Pull request by Francois-Gomis**
### Reason for change

[Linear task](https://linear.app/almapay/issue/ECOM-2370/[php-client]-add-v1mebusiness-event-endpoint)
Add business Event endpoints for analytics 

### Code changes

Add new DTO models  : 
OrderConfirmedBusinessEvent & CartInitiatedBusinessEvent

Add private endpoint and public methods to send DTO
sendOrderConfirmedBusinessEvent & sendCartInitiatedBusinessEvent

### How to test

_As a reviewer, you are encouraged to test the PR locally._

### Checklist for authors and reviewers

- [ ] The title of the PR uses business wording, not technical jargon, for the changelog readers to understand it
- [ ] The PR implements the changes asked in the referenced task / issue
- [ ] The automated tests are compliant with the [testing strategy](https://www.notion.so/almapay/Backend-testing-strategy-06c642cec1bf47b9b8feca3a91ea8d4a)
- [ ] The tests are relevant, and cover the corner/error cases, not only the happy path
- [ ] You understand the impact of this PR on existing code/features
- [ ] The changes include adequate logging and Datadog traces
- [ ] Documentation is updated (API, developer documentation, ADR, Notion...)

### CODE REVIEW

File: src/Endpoints/Merchants.php [hunk 1]
```diff
@@ -80,4 +83,61 @@ public function feePlans($kind = FeePlan::KIND_GENERAL, $installmentsCounts = "a
...
+    {
+        $cartEventDataPayload = [
+            'event_type' => $cartEventData->getEventType(),
+            'cart_id' => $cartEventData->getCartId()
+        ];
+        $this->sendBusinessEvent($cartEventDataPayload);
+    }
+
+    /**
+     * Prepare and send a business event for a cart initiated
```
Comment (joyet-simon) [inline src/Endpoints/Merchants.php:104]: Change the comment

File: tests/Unit/Endpoints/MerchantsTest.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        $this->requestObject->shouldReceive('setRequestBody')
+            ->once()
+            ->andReturn($this->requestObject);
+        $this->requestObject->shouldReceive('post')
+            ->once()
+            ->andThrow(new RequestError('Error in post', null, null));
+        $this->expectException(RequestException::class);
+        $this->merchantEndpoint->sendCartInitiatedBusinessEvent(new CartInitiatedBusinessEvent('42'));
+    }
+    public function testSendBusinessEventBadResponseRequestException()
```
Comment (joyet-simon) [inline tests/Unit/Endpoints/MerchantsTest.php:72]: add blank line

File: tests/Unit/Endpoints/MerchantsTest.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        $this->responseMock->shouldReceive('isError')
+            ->once()
+            ->andReturn(true);
+        $this->requestObject->shouldReceive('post')
+            ->once()
+            ->andReturn($this->responseMock);
+        $this->expectException(RequestException::class);
+        $this->merchantEndpoint->sendCartInitiatedBusinessEvent(new CartInitiatedBusinessEvent('42'));
+    }
+    /**
```
Comment (joyet-simon) [inline tests/Unit/Endpoints/MerchantsTest.php:91]: add blank line

File: src/Endpoints/Merchants.php [hunk 4]
```diff
@@ -80,4 +83,61 @@ public function feePlans($kind = FeePlan::KIND_GENERAL, $installmentsCounts = "a
...
+     * @param OrderConfirmedBusinessEvent $orderConfirmedBusinessEvent
+     * @return void
+     * @throws RequestException
+     */
+    public function sendOrderConfirmedBusinessEvent(OrderConfirmedBusinessEvent $orderConfirmedBusinessEvent)
+    {
+        $cartEventDataPayload = [
+            'event_type' => $orderConfirmedBusinessEvent->getEventType(),
+            'is_alma_p1x' => $orderConfirmedBusinessEvent->getIsAlmaP1X(),
+            'is_alma_bnpl' => $orderConfirmedBusinessEvent->getIsAlmaBNPL(),
```
Comment (webaaz) [inline src/Endpoints/Merchants.php:115]: same here

File: src/Endpoints/Merchants.php [hunk 5]
```diff
@@ -80,4 +83,61 @@ public function feePlans($kind = FeePlan::KIND_GENERAL, $installmentsCounts = "a
...
+            'was_bnpl_eligible' => $orderConfirmedBusinessEvent->getWasBNPLEligible(),
```
Comment (webaaz) [inline src/Endpoints/Merchants.php:116]: same here (why was?)

File: src/Entities/DTO/MerchantBusinessEvent/AbstractBusinessEvent.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+     * Get Event Type for merchant business event
+     *
+     * @return string
+     */
+    public function getEventType()
+    {
+        return $this->eventType;
+    }
+
+}
```
Comment (webaaz) [inline src/Entities/DTO/MerchantBusinessEvent/AbstractBusinessEvent.php:22]: Missing new line at the end of file

File: src/Entities/DTO/MerchantBusinessEvent/CartInitiatedBusinessEvent.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Get Cart Id
+     *
+     * @return string
+     */
+    public function getCartId()
+    {
+        return $this->cartId;
+    }
+}
```
Comment (webaaz) [inline src/Entities/DTO/MerchantBusinessEvent/CartInitiatedBusinessEvent.php:36]: same here

File: src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            !is_string($this->cartId) ||
+            // Alma payment id should be absent for non Alma payments
+            (!$this->isAlmaPayment() && !is_null($this->almaPaymentId))
+        )
+        {
+            throw new ParametersException('Invalid data type in OrderConfirmedBusinessEvent constructor');
+        }
+
+        //Alma payment id for Alma payment, Must be a string
+        if(
```
Comment (webaaz) [inline src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php:138]: this can be done on a single line, if it's two, the && must be at the start of the second line

File: tests/Unit/Endpoints/MerchantsTest.php [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+                    'alma_payment_id' => 'alma_payment_id'
+                ]
+            )
+            ->once()
+            ->andReturn($this->requestObject);
+        $this->responseMock->shouldReceive('isError')->once()->andReturn(false);
+        $this->requestObject->shouldReceive('post')->once()->andReturn($this->responseMock);
+        $this->assertNull($this->merchantEndpoint->sendOrderConfirmedBusinessEvent($orderConfirmedBusinessEvent));
+    }
+}
```
Comment (webaaz) [inline tests/Unit/Endpoints/MerchantsTest.php:173]: no new line at the end of file

File: tests/Unit/Entities/DTO/MerchantBusinessEvent/CartInitiatedBusinessEventTest.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            "cartId is an int" => [1],
+            "cartId is a float" => [1.1],
+            "cartId is an array" => [[]],
+            "cartId is an object" => [new \stdClass()],
+            "cartId is a boolean" => [true],
+            "cartId is null" => [null],
+        ];
+    }
+
+}
```
Comment (webaaz) [inline tests/Unit/Entities/DTO/MerchantBusinessEvent/CartInitiatedBusinessEventTest.php:41]: new line

File: tests/Unit/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEventTest.php [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+            "Cart id is a class" => [
+                'isP1X' => false,
+                'isBNPL' => false,
+                'wasEligible' => true,
+                'orderId' => '1',
+                'cartId' => new \stdClass()
+            ],
+        ];
+    }
+}
```
Comment (webaaz) [inline tests/Unit/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEventTest.php:276]: new line

**Review by webaaz — CHANGES_REQUESTED** [review]

File: src/Endpoints/Merchants.php (hunk 5 above)
Comment (Francois-Gomis) [reply src/Endpoints/Merchants.php:116]: was eligible at the time of payment

File: src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php (hunk 8 above)
Comment (Francois-Gomis) [reply src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php:138]: operator at end of line in project.

File: src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    protected function validateData()
+    {
+        if(
+            !is_bool($this->almaP1XStatus) ||
+            !is_bool($this->almaBNPLStatus) ||
+            !is_bool($this->wasBNPLEligible) ||
+            !is_string($this->orderId) ||
+            !is_string($this->cartId) ||
+            // Alma payment id should be absent for non Alma payments
+            (!$this->isAlmaPayment() && !is_null($this->almaPaymentId))
```
Comment (Benjamin-Freoua-Alma) [inline src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php:133]: Wouldn't it be better to put `!is_null` first because we check that the data is not null before checking this value?

**Review by Benjamin-Freoua-Alma — APPROVED** [review]

File: src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    /**
+     * For non alma payment, almaPaymentId should be null
+     * For Alma payment, almaPaymentId should be a string
+     *
+     * @param bool $isAlmaP1X
+     * @param bool $isAlmaBNPL
+     * @param bool $wasBNPLEligible
+     * @param string $orderId
+     * @param string $cartId
```
Comment (Benjamin-Freoua-Alma) [inline src/Entities/DTO/MerchantBusinessEvent/OrderConfirmedBusinessEvent.php:45]: add param almaPaymentId  docbloc

File: tests/Unit/Entities/DTO/MerchantBusinessEvent/CartInitiatedBusinessEventTest.php [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    public function testInvalidDataForBusinessEvent($cartId)
+    {
+        $this->expectException(ParametersException::class);
+        $this->expectExceptionMessage('CartId must be a string');
+        new CartInitiatedBusinessEvent($cartId);
+    }
+    public static function invalidDataForBusinessEventDataProvider()
+    {
+        return [
+            "cartId is an int" => [1],
```
Comment (Benjamin-Freoua-Alma) [inline tests/Unit/Entities/DTO/MerchantBusinessEvent/CartInitiatedBusinessEventTest.php:32]: add test empty string

**Review by Benjamin-Freoua-Alma — CHANGES_REQUESTED** [review]

**Review by Benjamin-Freoua-Alma — APPROVED** [review]
